File: Java/SinglyLLDeletion.java
//Deletion in Singly Linked List


import java.util.*;

class linkedList
{
    Node head;
    static class Node
    {
        int data;
        Node next;
        Node(int d)
        {
            data = d;
            next=null;
        }
    }

    static linkedList insertBeg(linkedList list, int val)
    {
        Node temp = new Node(val);
        temp.next = list.head;
        list.head = temp;
        return list;
    }

    static linkedList insertEnd(linkedList list, int val)
    {
        Node temp = new Node(val);
        if(list.head == null)
        {
            list.head= temp;
        }
        else
        {
            Node temp1 = list.head;
            while(temp1.next!=null)
            {
                temp1 = temp1.next;
            }
            temp1.next = temp;
        }
        return list;
        
    }

    static linkedList deleteBeg(linkedList list)
    {
        if(list.head == null)
        System.out.println("Not available");
        else
        {
            Node temp1 = list.head;
            temp1 = temp1.next;
            list.head = temp1;
        }
        return list;
    }

    static linkedList deleteEnd(linkedList list)
    {
        if(list.head == null)
        System.out.println("Not available");
        else if(list.head.next==null)
        {
            list.head = null;
            return list;
        }
        else
        {
            Node temp1 = list.head;
            while(temp1.next.next!=null)
            {
                temp1 = temp1.next;
            }
            temp1.next = null;

        }
        return list;
    }

    static void display(linkedList l)
    {
        Node temp= l.head;
        int sum =0,c=0;
        while(temp!=null)
        {
            System.out.print(temp.data + " ");
            sum += temp.data;
            temp = temp.next;
            c++;
        }
        System.out.println("Sum = " + sum + ", Average = " + (double)sum/c);
    }

    public static void main(String args[])
    {
        linkedList l = new linkedList();
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter total numbers to enter: ");
        int n = sc.nextInt();
        System.out.println("Select the following:\n1. Insert at beginning. \n2. Insert at end.");
        int cs = sc.nextInt();
        switch(cs)
        {
            case 1:
            {
                for(int i=0;i<n;i++)
                {
                    int val = sc.nextInt();
                    l = insertBeg(l,val);
                }
                break;
            }
            case 2:
            {
                for(int i=0;i<n;i++)
                {
                    int val = sc.nextInt();
                    l = insertEnd(l,val);
                }
                break;
            }
            default:
                System.out.println("Enter valid number");
                break;
        }
        
        // l = insertBeg(l,10);
        // l = insertBeg(l,20);
        // l= insertEnd(l,30);
        display(l);

        cs = sc.nextInt();
        switch(cs)
        {
            case 1:
                l = deleteBeg(l);
                break;
            
            case 2:    
                l = deleteEnd(l);
                break;
        }
        display(l);
        sc.close();
    }
}

File: C/SinglyLLDeletion.c
//Deletion in Singly Linked List


#include <stdio.h>
#include <stdlib.h>


//with malloc calloc -> use free
//with new operator use -> delete

struct ll 
{
    int data;
    struct ll *next;
};

struct ll *insertBegin(struct ll *p, int val)
{
    struct ll *temp;
    temp = (struct ll *)malloc(sizeof(struct ll));
    //temp = new struct linklist()  .....  in c++
    temp->data = val;
    temp->next = p;
    p = temp;
    return p;

}

struct ll *insertLast(struct ll *p,int val)
{
    struct ll *temp, *temp1;
    temp = (struct ll *)malloc(sizeof(struct ll));
    temp->data = val;
    temp->next = NULL;
    if(p==NULL)
    {
        p=temp;
    }
    else
    {
        temp1=p;
        while(temp1->next!=NULL)
        {
            temp1 = temp1->next;
        }
        temp1->next = temp;

    }
    return p;
    
}

void display(struct ll *p)
{
    struct ll *temp = p;
    while(temp!=NULL)
    {
        printf("%d ",temp->data);
        temp = temp->next;
    }
}

struct ll *deleteBeg(struct ll *p)
{
    struct ll *temp;
    if(p==NULL)
    {
        printf("Not available");
    }
    else
    {
        temp = p;
        p= p->next;
        free(temp);
    }
    return p;
    
}

struct ll *deleteEnd(struct ll *p)
{
    struct ll *temp, *s;
    if(p==NULL)
    {
        printf("Not available");
    }
    else
    {
        temp=p;
        while(temp->next->next!=NULL)
        {
            temp=temp->next;
        }
        s = temp->next;
        temp->next = NULL;
        
        free(s);
    }
    return p;
}


int main()
{
    struct ll *start;
    start = NULL;
    int n,val,c;
    printf("Enter total numbers to enter: ");
    scanf("%d",&n);
    printf("Select following: \n1. Insert at beginning. \n2. Insert at Last.\n");
    scanf("%d",&c);
    switch(c)
    {
        case 1: 
        {
            while(n-->0)
            {
                scanf("%d",&val);
                start = insertBegin(start,val);
            }
            break;
        }
        case 2:
        {
            while(n-->0)
            {
                scanf("%d",&val);
                start= insertLast(start,val);
            }
            break;
        }
        default:
        printf("Enter valid number");
        break;
    }
    
    display(start);

    printf("Selct following: \n1. Delete from beginning. \n2. Delete from end.");
    scanf("%d",&c);
    switch(c)
    {
        case 1: 
            start = deleteBeg(start);
             break;
        case 2:
            start = deleteEnd(start);
            break;
        default:
            printf("Enter valid number");
            break;
    }
    display(start);

}

